### ISSUE TITLE
Query watcher not being called when cache is updated on an element by another query/subscription/mutation

### ISSUE THREAD

**Issue by benoitletondor** [apollographql/apollo-ios#1422]
## Bug report

Query watcher not being called when cache is updated on an element of a collection that is added after calling watch

## Versions

Please fill in the versions you're currently using: 

- `apollo-ios` SDK version: 0.34.0
- Xcode version: 12.0.1
- Swift version: 5.3

## Steps to reproduce

I've been trying a lot of things on that one and wasn't able to find a way to fix my problem. So in my app I'm having a list of conversations, so I'm creating a `watch` on a query that returns the list of conversations, this is something like that:

```
user {
    uuid
    firstName
    lastName
    conversations {
        uuid
        unreadMessagesCount
    }
}
```

> In the app we use the `uuid` key to handle cache, so we make sure to always pass `uuid` in our queries to automatically handle cache update.

So the role of the watcher I'm talking about is both to update existing conversations and also be able to catch when a new conversation is created (it can happen and not be initiated by the user, we then trigger an event from the backend that is listen from a subscription in the app). When this event happens it returns something like that

```
event {
     newConversation {
          conversation {
               uuid
               unreadMessagesCount
               // This is the part that adds the new conversation to the existing ones of the users in the cache
               user {
                    uuid
                    conversations {
                          uuid
                          unreadMessagesCount
                    }
               } 
          }
     }
}
```

It works great, meaning that when this event happens, the watcher is being called with the newly created conversation, but the issue is that any new cache update for that specific conversation doesn't trigger the watcher again.

After investigating a bit, I realised that the cache is being updated because if I'm adding 3 new messages into the new conversation (setting the `unreadMessagesCount` to 3), the watcher doesn't get called but then if I add 1 new message into an old one, the watcher is being called with both the new message on the old conversation and the 3 new on the new one. 

So it really seems like `watch` is not being called again for changes on an item that wasn't in a collection when the `watch` was initially made. I've also take a look at https://github.com/apollographql/apollo-ios/issues/281 and making a `fetch` on the same query again after the event doesn't fix the issue.

Let me know if I'm not clear on something as the whole thing is a bit complicated to explain.

**Comment by benoitletondor** [comment]
To add more to this, I feel like this is not related to collections only. If you have 1 query that watches something, then another query/mutation updates this cache indirectly (not using the same query, but another query that contains the same element with the same `uuid`), it doesn't work.

**Comment by designatednerd** [comment]
Do you have any thoughts on why the test that was added would be passing, but what you're working with would not be working?

**Comment by benoitletondor** [comment]
I'm afraid no, but all I can tell is that it works with 0.32.0 and not with 0.34.0

**Comment by Narayane** [comment]
I think I have the same problem but not on a collection. My watcher is not triggered anymore since I have updated Apollo from 0.33.0 to 0.34.1. I have a watcher in cachePolicy .returnCacheDataAndFetch on this Query.

```
query GetRide($rideId: ID!) {
    ride(id: $rideId) {
        ...EndRideFragment
    }
}
```

And I have a mutation that updates my cache and normally triggered my watcher.
```
mutation StopRide($input: StopActiveRideInput!) {
    stopActiveRide(input: $input) {
        ride {
            ...EndRideFragment
        }
    }
}
```
Since my Apollo pod upgrade, not working anymore. My watcher is not triggered when mutation payload updates my cache.

**Comment by dhritzkiv** [comment]
This is also occurring for me. 0.32.1 worked; but not 0.34.0 or 0.34.1

**Comment by designatednerd** [comment]
@dhritzkiv That makes sense, the network stack and how it interacts with the cache changed completely in 0.34.0. 

Apologies, I've been under the weather. I'm gonna do some digging starting tomorrow to try to figure out how I broke this without breaking the tests 🙃

**Comment by dhritzkiv** [comment]
No rush from me! I was able to roll back to 0.33.0 no problem

**Comment by Narayane** [comment]
I can wait 1 week to have feedback from you @designatednerd on this point.
If you need more time to investigate or to solve it, I could roll back to 0.33.0 to avoid blocking me in production

**Comment by benoitletondor** [comment]
Same here, I rollbacked to 0.32.1 so no rush from me, and thanks for taking time investigating this!

**Comment by martijnwalraven** [comment]
I'm trying to reproduce this issue to investigate what could be going on, but it's been hard because unfortunately our tests aren't set up to run in real life conditions, where concurrency may play a role. The existing watcher tests all continue to pass, and so do some new ones I've been adding.

It would be helpful to learn more about the way people who experience this are using the framework, and what behavior they are seeing:
- Does the issue occur 100% of the time, or only sometimes?
- Are there other operations in flight when this happens, or does it also happen with just a single watcher and a single related fetch?
- Are you calling methods on `ApolloClient` from the main thread or from a background thread? Do you receive results on the main queue as well, or are you passing in a custom queue?
- Anything else unusual or noticeable?

I know this is a long shot, but it would be even more helpful if anyone was able to share a project that reliably exhibits this issue.

**Comment by martijnwalraven** [comment]
A few more thoughts after taking a closer look at this:

1) It seems any operations using the `WebSocketTransport` always ignore the cache completely starting in 0.33 (I believe https://github.com/apollographql/apollo-ios/commit/444c465671c786659599a16da4b2d23d43648a74 is where this change was introduced). That means subscription results are not actually published to the store, and thus will not trigger watchers. (@benoitletondor I think that would at least explain your initial bug report).

2) That doesn't yet explain why queries and mutations would also fail to trigger watchers (assuming those are not using the `WebSocketTransport`). There is a possibility for misconfiguration however, which would lead to inadvertently having multiple stores (see https://github.com/apollographql/apollo-ios/issues/1438). Could some of you be bitten by this maybe?

**Comment by dhritzkiv** [comment]
I'm not using WebSocketTransport, but 3. sound like it could be what's plaguing us! Will have some time in two weeks to poke around that

**Comment by Narayane** [comment]
@martijnwalraven I guess 3. could be an explanation for my case, I have a custom `NetworkTransport` implementation

**Comment by Narayane** [comment]
@designatednerd @martijnwalraven any progress?

**Comment by designatednerd** [comment]
not yet -  also, be aware that `parseResultFast` doesn't hit the cache, so you'd probably need to use the completion-closure based parsing.

**Comment by Narayane** [comment]
What?
A code example could save my day :)

**Comment by designatednerd** [comment]
@Narayane @benoitletondor @dhritzkiv I *think* this should be resolved by `0.37.0`. Could y'all give that a try and let me know if it's still an issue on that version?

**Comment by dhritzkiv** [comment]
@designatednerd A preliminary run using 0.37.0: it looks like the cache+watchers work as expected! Thanks. :)

**Comment by Narayane** [comment]
@designatednerd Still an issue for me, watchers are not triggered in 0.37.0 with the code below in my custom `NetworkTransport`.

```
do {
    let body = try self.serializationFormat.deserialize(data: data) as! JSONObject
    let response = GraphQLResponse(operation: operation, body: body)
    response.parseResultWithCompletion(completion: { result in
         switch result {
         case .success(let gqlResult):
              completionHandler(.success(gqlResult.0))
         case .failure(let error):
              completionHandler(.failure(LRError.clientError(error.localizedDescription)))
         }
     })
} catch { }
```

**Comment by designatednerd** [comment]
@Narayane Yeah, the caching on custom network transports is basically broken at the moment, so that would make sense. The current workaround is to write a custom interceptor for your network usage similar to [the default Network interceptor](https://github.com/apollographql/apollo-ios/blob/main/Sources/Apollo/NetworkFetchInterceptor.swift) but using your own networking. 

I'm working on a longer-term fix but this is something that was missed through the RFC and Beta processes so there's a decent amount of rethinking that needs to be done there, unfortunately 😭

**Comment by Narayane** [comment]
Hi,

any progress on my problem?

It begins to be disturbing to be stuck in v0.33 since v0.39 is the current...

**Comment by designatednerd** [comment]
At this point the recommended workaround remains creating your own custom network interceptor and using the `RequestChainNetworkTransport` rather than creating your own network transport - this should work in almost all cases, I would love to hear any cases where it doesn't so I can fix whatever needs to be fixed to make that work.

This is definitely something that needs to be revisited but at this point i think the workaround is sufficient that it'll be better long-term to do it as part of enhancements to the chain system that will make it easier to use this stuff for all network requests (including, hopefully, web sockets).

**Comment by Narayane** [comment]
@designatednerd Ok, I don't have time for the moment but I would try.

I have another problem with last Xcode version 12.4 now in function signature which don't have with previous.

![Capture d’écran 2021-01-29 à 12 47 20](https://user-images.githubusercontent.com/4921308/106271925-812b8f80-6230-11eb-90eb-9094e84e4433.png)

Could you help me?

**Comment by ketenshi** [comment]
Hi all,

Is there a recommended way to deal with the subscriptions from the `WebSocketTransport` not publishing to the store?

**Comment by designatednerd** [comment]
@Narayane I don't have an example, but if you look at how the default network intercept provider is set up, you'll see it's fairly simple. 

@ketenshi That's a more complicated issue, unfortunately - we don't presently have a recommended workaround but i'm working on some thoughts around it.

**Comment by Narayane** [comment]
@designatednerd

https://github.com/apollographql/apollo-ios/blob/main/Sources/Apollo/NetworkFetchInterceptor.swift

```
switch result {
      case .failure(let error):
      case .success(let (data, httpResponse)):
}
```

In my previous implementation of `NetworkTransport`, I could make out cases like _not an httpResponse_ (client or server timeout), _no data_ and all httpResponse status codes to determine precisely what kind of error it was.
I guess I can make out anymore now, with this example, not an httpResponse and no data cases, right? If yes, it is a true problem to me because I need to retry in some specific cases considering error type returned... Can you enlighten me?

**Comment by designatednerd** [comment]
If you're using our `URLSessionClient`, it will return [a specific error](https://github.com/apollographql/apollo-ios/blob/b89f7c309adb72e75c17d5f8d80970ade2e0d168/Sources/Apollo/URLSessionClient.swift#L17) if no `HTTPURLResponse` is returned (the `TaskData` object does the work of [checking if it's an `HTTPURLResponse`](https://github.com/apollographql/apollo-ios/blob/b89f7c309adb72e75c17d5f8d80970ade2e0d168/Sources/Apollo/TaskData.swift#L21). 

`Data` will always be returned since we need to [instantiate the data to keep appending to it on the `TaskData`](https://github.com/apollographql/apollo-ios/blob/b89f7c309adb72e75c17d5f8d80970ade2e0d168/Sources/Apollo/TaskData.swift#L8), but you can check if it's empty. 

`HTTPURLResponse` has a `statusCode` property you can check for the exact HTTP status code returned. 

Does that help?

**Comment by Narayane** [comment]
I don't know if I am currently using an `URLSessionClient`. In my code, I just instantiate an `ApolloClient` with an `ApolloStore` and a custom `NetworkTransport`.

```
self.transport = LRNetworkTransport(url: endpoint)
do {
    let cache = try SQLiteNormalizedCache(fileURL: cacheFilePath)
    self.store = ApolloStore(cache: cache)
    self.client = ApolloClient(networkTransport: self.transport, store: self.store!)
} catch {}
```

Your other answers helps me.

**Comment by designatednerd** [comment]
You are not if you're providing the alternate network transport.

**Comment by Narayane** [comment]
Is it ok if I follow [this](https://www.apollographql.com/docs/ios/initialization/#example-network-singleton-setup) to make evolve the way I build my custom Apollo Client for resolving my watchers trigger problems with v0.34+?

**Comment by designatednerd** [comment]
I'm not quite sure what the question is here - that is currently the recommended setup. Can you clarify a bit what the question is?

**Comment by benoitletondor** [comment]
@designatednerd 

Taking the occasion to jump again on the conversation. Please let me know when you have news around the `WebSocketTransport` not publishing into the store bug, this is the last one preventing us from using the latest versions.

Many thanks!

**Comment by designatednerd** [comment]
Planning on tackling this soon!

**Comment by tomasyany** [comment]
@designatednerd Do you have any updates on this issue?

**Comment by designatednerd** [comment]
Thanks for the poke! Right now the main update is that we've had to roll back Starscream to an older version due to some crash issues a bunch of people were experiencing. 

We're looking at how to address this hole (and others in the current request chain implementation) in a more long-term way right now, but I don't have an ETA. 

Apologies for the back and forth on this. We've got another dev who's getting ramped up now, so I'm hoping we'll have a much better process and predictability around all this soon.
